Functional tests for nova-compute 'cloud' actions

mkalcok opened the pull request:
New actions were added to the nova-compute unit:

*  Actions that can `enable` or `disable` scheduling of new VMs on the nova-compute.
* Actions that can register/unregister nova-compute from nova-cloud controller.

This PR contains functional test that tests these four actions.

nova-compute code review: https://review.opendev.org/c/openstack/charm-nova-compute/+/763795

auria commented:
LGTM.

ajkavanagh reviewed (changes requested):
Please could you provide a link to the gerrit review for the change, so we can see the matching code in the nova-compute charm?  Also, please could you describe the change in the PR header for future travelers.  Thanks.  (Please also see inline comment).

mkalcok commented:
@ajkavanagh Sorry about that description. It should be ok now. However I don't see the inline comments you mention

ajkavanagh commented:
> @ajkavanagh Sorry about that description. It should be ok now. However I don't see the inline comments you mention

It looks like my comment went missing (probably due to an error I made!).  I'll re-do it.

ajkavanagh commented on zaza/openstack/charm_tests/nova/tests.py:
```diff
@@ -154,6 +154,31 @@ def test_930_check_virsh_default_network(self):
                 model_name=self.model_name)
             self.assertFalse(int(run['Code']) == 0)
 
+    def test_940_enable_disable_actions(self):
+        """Test disable/enable actions on nova-compute units."""
+        nova_units = zaza.model.get_units('nova-compute',
+                                          model_name=self.model_name)
+
+        # Check that nova-compute services are enabled before testing
+        for service in self.nova_client.services.list(binary='nova-compute'):
+            self.assertEqual(service.status, 'enabled')
+
+        # Run 'disable' action on units
+        zaza.model.run_action_on_units([unit.name for unit in nova_units],
+                                       'disable')
+
+        # Check action results via nova API
+        for service in self.nova_client.services.list(binary='nova-compute'):
+            self.assertEqual(service.status, 'disabled')
+
+        # Run 'enable' action on units
+        zaza.model.run_action_on_units([unit.name for unit in nova_units],
+                                       'enable')
+
+        # Check action results via nova API
+        for service in self.nova_client.services.list(binary='nova-compute'):
+            self.assertEqual(service.status, 'enabled')
+
```
So if this is a test for a new feature, this will break on old versions of the charm, specifically the stable series.  As we don't (yet) cut stable branches of these tests, it's normal practice to add a new test class with this test (for a new feature), and add the test to the tests/tests.yaml for the master branch of the charm.  Then, the new version gets the new tests, and the stable version of the charm does.  This means that we will continue to be able to backport changes to the stable series.

mkalcok replied:
Good point, thanks. I forgot that this was the procedure for new func tests.

mkalcok replied:
done

ajkavanagh reviewed (commented):
Looks good; let's see the associated gerrit review pass prior to landing and it'll be good to merge.

mkalcok commented:
I added functional tests for actions `remove-from-cloud` and `register-to-cloud`. With their addition I also renamed class that encapsulate theses tests to `CloudActions` as it now covers all actions from nova-compute's `actions/cloud.py`

ajkavanagh reviewed (changes requested):
Holding until charm-freeze is over; don't want to disturb the tests during the testing week.

mkalcok commented:
@ajkavanagh Hi Alex, the associated gerrit review went through. https://review.opendev.org/c/openstack/charm-nova-compute/+/763795

ajkavanagh commented on zaza/openstack/charm_tests/nova/tests.py:
```diff
@@ -68,6 +78,139 @@ def test_launch_small_instance(self):
+        # nova-cloud-controller
+        sleep_timeout = 1  # don't waste 10 seconds on the first run
+
+        for _ in range(31):
+            sleep(sleep_timeout)
+            service_list = self.nova_client.services.list(
+                host=service_name, binary='nova-compute')
+            if len(service_list) == 0:
+                break
+            sleep_timeout = 10
+        else:
+            self.fail("nova-compute service was not unregistered from the "
+                      "nova-cloud-controller as expected.")
```
This code is sadly repeated below, which violates DRY.  There are several ways to make it better; I suspect tenacity would be a good approach.  However, sadly the actual review in the charm has been landed so this PR needs to be landed.  We'll have to fix it later.

ajkavanagh reviewed (approved):
Approving despite reservation on repeated code; the associated gerrit change the charm has already landed, which is generally the wrong order.

mkalcok commented:
Thanks for merge @ajkavanagh . You are right, the repeated code could've been handled better. I opened PR #534 that should solve it.

ajkavanagh commented:
> Thanks for the merge @ajkavanagh . You are right, the repeated code could've been handled better. I opened PR #534 that should solve it.

Yes, thank you; seen that and merged it.  Incidentally, it wasn't your 'problem'; sorry if it seemed like that.  I was more sad that the gerrit review had been landed before the associated tests had been fully reviewed and merged.  It kind of forces the issue! :)
